File: apps/frontend/src/types/wedding.ts
// Semua interface di sini di-selaraskan manual dengan struct Rust di
// apps/backend/src/models/wedding.rs dan skema di apps/backend/migrations/*.sql.
// Kalau salah satu berubah, update juga sisi satunya.

// ---------------------------------------------------------------------------
// Baris tabel Postgres mentah (belum ada endpoint yang mengembalikan bentuk ini
// apa adanya, tapi berguna sebagai referensi tipe kalau nanti dibutuhkan --
// mis. admin table view langsung dari DB).
// ---------------------------------------------------------------------------

/** Mirror tabel `weddings` -- migrations/0001_create_weddings.sql */
export interface Wedding {
  id: string;
  subdomain_slug: string;
  theme_id: number;
  primary_color: string;
  secondary_color: string;
  music_url: string | null;
  access_token: string;
  is_published: boolean;
  created_at: string;
}

/** Mirror tabel `wedding_details` -- migrations/0002_create_wedding_details.sql */
export interface WeddingDetail {
  id: string;
  wedding_id: string;
  groom_name: string;
  bride_name: string;
  wedding_date: string | null;
  location_address: string | null;
  maps_url: string | null;
}

/** Mirror tabel `rsvp` -- migrations/0003_create_rsvp.sql (belum ada endpoint API-nya) */
export type AttendanceStatus = 'attending' | 'not_attending' | 'maybe';

export interface RsvpData {
  id: string;
  wedding_id: string;
  guest_name: string;
  attendance_status: AttendanceStatus;
  message: string | null;
  created_at: string;
}

// ---------------------------------------------------------------------------
// Bentuk JSON asli response backend (Rust: models/wedding.rs) -- INI yang
// benar-benar dikonsumsi oleh fetch() di frontend.
// ---------------------------------------------------------------------------

/** Rust: ThemeDto */
export interface ThemeInfo {
  theme_id: number;
  primary_color: string;
  secondary_color: string;
}

/** Rust: CoupleDto */
export interface CoupleInfo {
  groom_name: string;
  bride_name: string;
  groom_photo_url?: string | null;
  bride_photo_url?: string | null;
  groom_parents?: string | null;
  bride_parents?: string | null;
  groom_ig?: string | null;
  bride_ig?: string | null;
}

/** Rust: EventDto */
export interface EventInfo {
  wedding_date: string | null;
  location_address: string | null;
  maps_url: string | null;
  akad_date?: string | null;
  akad_location?: string | null;
  akad_maps_url?: string | null;
  resepsi_date?: string | null;
  resepsi_location?: string | null;
  resepsi_maps_url?: string | null;
}

export interface LoveStoryInfo {
  date: string | null;
  description: string | null;
  photo_url: string | null;
}

export interface GalleryPhotoInfo {
  photo_url: string;
}

export interface WeddingGiftInfo {
  gift_type: string;
  bank_name: string | null;
  account_name: string | null;
  account_number: string | null;
}

/**
 * Rust: ContactSettingsDto -- setting GLOBAL platform (tabel `platform_settings`,
 * satu baris untuk semua wedding, diatur lewat Admin Dashboard), bukan data per-wedding.
 */
export interface ContactSettings {
  contact_instagram_url: string | null;
  contact_whatsapp_url: string | null;
  contact_handle: string | null;
}

/** Rust: WeddingDetailsResponse -- response GET /api/wedding-details */
export interface WeddingData {
  subdomain: string;
  is_published: boolean;
  theme: ThemeInfo;
  couple: CoupleInfo;
  event: EventInfo;
  music_url: string | null;
  gallery_video_url?: string | null;
  love_stories: LoveStoryInfo[];
  gallery_photos: GalleryPhotoInfo[];
  wedding_gifts: WeddingGiftInfo[];
  theme_settings?: Record<string, any> | null;
  cover_photo_url?: string | null;
  contact: ContactSettings;
}

/** Rust: CreateWeddingRequest -- payload POST /api/weddings */
export interface CreateWeddingPayload {
  groom_name: string;
  bride_name: string;
  theme_id?: number;
}

/** Rust: CreateWeddingResponse -- response POST /api/weddings */
export interface CreateWeddingResponse {
  wedding_id: string;
  subdomain_slug: string;
  access_token: string;
  groom_name: string;
  bride_name: string;
  theme_id: number;
  is_published: boolean;
}

/** Props yang wajib diterima setiap komponen tema (Theme1, Theme2, dst). */
export interface ThemeComponentProps {
  data: WeddingData;
  guestName?: string;
}

// ---------------------------------------------------------------------------
// Editor berbasis token (Langkah 5) -- Rust: models/wedding_edit.rs
// ---------------------------------------------------------------------------

/** Rust: LoveStoryEntry -- satu baris Love Story di editor (dua arah). */
export interface LoveStoryEntry {
  date: string | null;
  description: string | null;
  photo_url: string | null;
}

/** Rust: WeddingGiftEntry -- satu baris Wedding Gift di editor (dua arah). */
export interface WeddingGiftEntry {
  /** "bank" (rekening) atau "kado" (alamat kirim kado). */
  gift_type: string;
  bank_name: string | null;
  account_name: string | null;
  account_number: string | null;
}

/** Rust: WeddingEditData -- response GET /api/wedding/edit-auth & PUT /api/wedding/update */
export interface WeddingEditData {
  wedding_id: string;
  subdomain_slug: string;
  theme_id: number;
  primary_color: string;
  secondary_color: string;
  music_url: string | null;
  groom_name: string;
  bride_name: string;
  wedding_date: string | null;
  location_address: string | null;
  maps_url: string | null;
  cover_photo_url: string | null;
  groom_photo_url: string | null;
  bride_photo_url: string | null;
  groom_parents: string | null;
  bride_parents: string | null;
  groom_ig: string | null;
  bride_ig: string | null;
  akad_date: string | null;
  akad_location: string | null;
  akad_maps_url: string | null;
  resepsi_date: string | null;
  resepsi_location: string | null;
  resepsi_maps_url: string | null;
  gallery_video_url: string | null;
  theme_settings?: Record<string, any> | null;
  /** Versi data untuk optimistic locking -- kirim balik sebagai expected_updated_at. */
  updated_at: string;
  love_stories: LoveStoryEntry[];
  gallery_photos: string[];
  wedding_gifts: WeddingGiftEntry[];
}

/** Rust: UpdateWeddingPayload -- payload PUT /api/wedding/update.
 *  Array bersifat replace-all: isi lama diganti persis dengan isi payload. */
export interface UpdateWeddingPayload {
  groom_name: string;
  bride_name: string;
  primary_color: string;
  secondary_color: string;
  wedding_date: string | null;
  location_address: string | null;
  maps_url: string | null;
  cover_photo_url: string | null;
  /** file_url salah satu baris music_library, diisi lewat picker (bukan upload bebas). */
  music_url: string | null;
  groom_photo_url: string | null;
  bride_photo_url: string | null;
  groom_parents: string | null;
  bride_parents: string | null;
  groom_ig: string | null;
  bride_ig: string | null;
  akad_date: string | null;
  akad_location: string | null;
  akad_maps_url: string | null;
  resepsi_date: string | null;
  resepsi_location: string | null;
  resepsi_maps_url: string | null;
  gallery_video_url: string | null;
  theme_settings?: Record<string, any> | null;
  /** updated_at yang diterima saat memuat form; backend menolak 409 kalau
   *  nilainya sudah tidak cocok (data diubah dari tempat lain). */
  expected_updated_at: string | null;
  love_stories: LoveStoryEntry[];
  gallery_photos: string[];
  wedding_gifts: WeddingGiftEntry[];
}

/** Rust: UploadResponse -- response POST /api/wedding/upload */
export interface UploadResponse {
  url: string;
}

// ---------------------------------------------------------------------------
// Admin login -- Rust: models/admin.rs
// ---------------------------------------------------------------------------

/** Rust: AdminLoginPayload -- payload POST /api/admin/login */
export interface AdminLoginPayload {
  username: string;
  password: string;
}

/** Rust: AdminLoginResponse -- response POST /api/admin/login */
export interface AdminLoginResponse {
  success: boolean;
}

/**
 * Rust: MusicTrackDto -- satu baris tabel `music_library`. Diupload sekali oleh
 * admin, dipakai berkali-kali oleh banyak wedding lewat picker di editor.
 */
export interface MusicTrack {
  id: string;
  title: string;
  artist: string | null;
  file_url: string;
}

/** Rust: WeddingSummaryDto -- satu baris monitoring di GET /api/admin/weddings */
export interface WeddingSummary {
  subdomain_slug: string;
  /** Ditampilkan supaya admin bisa salin ulang kalau pasangan lupa token editornya. */
  access_token: string;
  groom_name: string;
  bride_name: string;
  theme_id: number;
  is_published: boolean;
  /** Masa aktif (ISO); null = tanpa batas. Lewat tanggal ini tamu dapat 404. */
  active_until: string | null;
  created_at: string;
  rsvp_count: number;
}

/** Rust: WeddingListResponse -- response GET /api/admin/weddings (berhalaman) */
export interface WeddingListResponse {
  weddings: WeddingSummary[];
  total: number;
  page: number;
  per_page: number;
}

/** Rust: UpdateSettingsPayload -- payload PUT /api/admin/settings */
export interface UpdateSettingsPayload {
  contact_instagram_url: string | null;
  contact_whatsapp_url: string | null;
  contact_handle: string | null;
}

File: apps/backend/src/models/wedding.rs
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use uuid::Uuid;

// ---------- POST /api/weddings ----------

#[derive(Debug, Deserialize)]
pub struct CreateWeddingRequest {
    pub groom_name: String,
    pub bride_name: String,
    pub theme_id: Option<i32>,
}

#[derive(Debug, Serialize)]
pub struct CreateWeddingResponse {
    pub wedding_id: Uuid,
    pub subdomain_slug: String,
    pub access_token: String,
    pub groom_name: String,
    pub bride_name: String,
    pub theme_id: i32,
    pub is_published: bool,
}

// ---------- GET /api/wedding-details ----------

/// Hasil JOIN mentah dari tabel `weddings` + `wedding_details`.
#[derive(Debug, sqlx::FromRow)]
pub struct WeddingDetailsRow {
    pub subdomain_slug: String,
    pub theme_id: i32,
    pub primary_color: String,
    pub secondary_color: String,
    pub music_url: Option<String>,
    pub is_published: bool,
    /// Masa aktif; NULL = tanpa batas. Lewat tanggal ini undangan diperlakukan
    /// seperti draft (tamu 404, pemilik ber-token tetap bisa lihat).
    pub active_until: Option<DateTime<Utc>>,
    /// Dipakai handler untuk mengizinkan pasangan melihat draft-nya sendiri
    /// (header X-Access-Token) -- TIDAK ikut diserialisasi ke response publik.
    pub access_token: String,
    pub groom_name: String,
    pub bride_name: String,
    pub wedding_date: Option<DateTime<Utc>>,
    pub location_address: Option<String>,
    pub maps_url: Option<String>,
    pub cover_photo_url: Option<String>,
    pub groom_photo_url: Option<String>,
    pub bride_photo_url: Option<String>,
    pub groom_parents: Option<String>,
    pub bride_parents: Option<String>,
    pub groom_ig: Option<String>,
    pub bride_ig: Option<String>,
    pub akad_date: Option<DateTime<Utc>>,
    pub akad_location: Option<String>,
    pub akad_maps_url: Option<String>,
    pub resepsi_date: Option<DateTime<Utc>>,
    pub resepsi_location: Option<String>,
    pub resepsi_maps_url: Option<String>,
    pub gallery_video_url: Option<String>,
    pub theme_settings: Option<serde_json::Value>,
}

#[derive(Debug, Serialize)]
pub struct WeddingDetailsResponse {
    pub subdomain: String,
    pub is_published: bool,
    pub theme: ThemeDto,
    pub couple: CoupleDto,
    pub event: EventDto,
    pub music_url: Option<String>,
    pub gallery_video_url: Option<String>,
    pub love_stories: Vec<LoveStoryDto>,
    pub gallery_photos: Vec<GalleryPhotoDto>,
    pub wedding_gifts: Vec<WeddingGiftDto>,
    pub theme_settings: Option<serde_json::Value>,
    pub cover_photo_url: Option<String>,
    pub contact: ContactSettingsDto,
}

/// Setting GLOBAL platform (tabel `platform_settings`, satu baris untuk semua
/// wedding) -- dipakai footer "Hubungi Kami", diatur lewat Admin Dashboard.
#[derive(Debug, Default, Serialize, sqlx::FromRow)]
pub struct ContactSettingsDto {
    pub contact_instagram_url: Option<String>,
    pub contact_whatsapp_url: Option<String>,
    pub contact_handle: Option<String>,
}

#[derive(Debug, Serialize)]
pub struct ThemeDto {
    pub theme_id: i32,
    pub primary_color: String,
    pub secondary_color: String,
}

#[derive(Debug, Serialize)]
pub struct CoupleDto {
    pub groom_name: String,
    pub bride_name: String,
    pub groom_photo_url: Option<String>,
    pub bride_photo_url: Option<String>,
    pub groom_parents: Option<String>,
    pub bride_parents: Option<String>,
    pub groom_ig: Option<String>,
    pub bride_ig: Option<String>,
}

#[derive(Debug, Serialize)]
pub struct EventDto {
    pub wedding_date: Option<DateTime<Utc>>,
    pub location_address: Option<String>,
    pub maps_url: Option<String>,
    pub akad_date: Option<DateTime<Utc>>,
    pub akad_location: Option<String>,
    pub akad_maps_url: Option<String>,
    pub resepsi_date: Option<DateTime<Utc>>,
    pub resepsi_location: Option<String>,
    pub resepsi_maps_url: Option<String>,
}

#[derive(Debug, Serialize, sqlx::FromRow)]
pub struct LoveStoryDto {
    pub date: Option<String>,
    pub description: Option<String>,
    pub photo_url: Option<String>,
}

#[derive(Debug, Serialize, sqlx::FromRow)]
pub struct GalleryPhotoDto {
    pub photo_url: String,
}

#[derive(Debug, Serialize, sqlx::FromRow)]
pub struct WeddingGiftDto {
    pub gift_type: String,
    pub bank_name: Option<String>,
    pub account_name: Option<String>,
    pub account_number: Option<String>,
}

impl From<WeddingDetailsRow> for WeddingDetailsResponse {
    fn from(row: WeddingDetailsRow) -> Self {
        Self {
            subdomain: row.subdomain_slug,
            is_published: row.is_published,
            theme: ThemeDto {
                theme_id: row.theme_id,
                primary_color: row.primary_color,
                secondary_color: row.secondary_color,
            },
            couple: CoupleDto {
                groom_name: row.groom_name,
                bride_name: row.bride_name,
                groom_photo_url: row.groom_photo_url,
                bride_photo_url: row.bride_photo_url,
                groom_parents: row.groom_parents,
                bride_parents: row.bride_parents,
                groom_ig: row.groom_ig,
                bride_ig: row.bride_ig,
            },
            event: EventDto {
                wedding_date: row.wedding_date,
                location_address: row.location_address,
                maps_url: row.maps_url,
                akad_date: row.akad_date,
                akad_location: row.akad_location,
                akad_maps_url: row.akad_maps_url,
                resepsi_date: row.resepsi_date,
                resepsi_location: row.resepsi_location,
                resepsi_maps_url: row.resepsi_maps_url,
            },
            music_url: row.music_url,
            gallery_video_url: row.gallery_video_url,
            love_stories: vec![],
            gallery_photos: vec![],
            wedding_gifts: vec![],
            theme_settings: row.theme_settings,
            cover_photo_url: row.cover_photo_url,
            contact: ContactSettingsDto::default(),
        }
    }
}
